lib/portlayer/exec/commit_test.go

```diff
@@ -0,0 +1,109 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package exec
+
+import (
+	"context"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	
+	"github.com/vmware/vic/pkg/trace"
+)
+
+func TestBatchBlockOnFunc(t *testing.T) {
+	ctx := context.Background()
+	operation := func(op trace.Operation) error {
+		time.Sleep(2 * time.Second)
+		return nil
+	}
+
+	// test fetch one request
+	testFetchOne(t, ctx, operation)
+
+	// serialize test: batch size 1, 3 requests
+	// batch size 1 = no batching. Process requests one at a time
+	// total time spent should be within: (# of requests) * (operation wait time)
+	testMultipleBatch(t, ctx, 1, 3, operation)
+
+	// batch size 10, 5 requests
+	// the 5 requests should be processed in one batch
+	// total time spent should be within: 1 * (operation wait time)
+	testMultipleBatch(t, ctx, 10, 5, operation)
+
+	//  batch size 100, 50 requests
+	// the 50 requests should be processed in one batch
+	// total time spent should be within: 1 * (operation wait time)
+	testMultipleBatch(t, ctx, 100, 50, operation)
+
+	// stress test: batch size 100, 200 requests
+	// the 200 requests should be processed in 2 batches
+	// total time spent should be within: 2 * (operation wait time)
+	testMultipleBatch(t, ctx, 100, 200, operation)
+
+	// stress test: batch size 100, 500 requests
+	// the 500 requests should be processed in 5 batches
+	// total time spent should be within: 5 * (operation wait time)
+	testMultipleBatch(t, ctx, 100, 500, operation)
+}
+
+func testFetchOne(t *testing.T, ctx context.Context, operation func(op trace.Operation) error) {
+	batch := make(chan chan error, 5) // batch size 5
+
+	// fire background reader
+	go batchBlockOnFunc(ctx, batch, operation)
+
+	// send only 1 request
+	sendRequest(t, batch, nil)
+	close(batch)
+}
+
+func testMultipleBatch(t *testing.T, ctx context.Context, batchSize int, totalRequest int, operation func(op trace.Operation) error) {
+	batch := make(chan chan error, batchSize)
+
+	done := make([]chan bool, totalRequest)
+	for i := 0; i < totalRequest; i++ {
+		done[i] = make(chan bool, 1)
+	}
+
+	// fire background request reader
+	go batchBlockOnFunc(ctx, batch, operation)
+
+	// send requests concurrently
+	for i := 0; i < totalRequest; i++ {
+		go sendRequest(t, batch, done[i])
+	}
+
+	// wait until all requests are processed, close batch channel and quit background receiver
+	quitBatchUntilDone(done, batch)
```

Comment:
We never actually verify that the operation is run.

If I replace line 261 of `commit.go` (`err := operation(op)`) with `var err error` (i.e., remove the call to `operation`), the tests still pass.

In addition to sleeping, it seems like we need the operation to _do_ something that we can make an assertion about (e.g., send something to a channel or increment a counter).

Reply:
Looking at the code, I don't see an actual limit on the batch size: just a maximum capacity for the channel. (And because the channel is being drained as it is being added to, we can end up with more than the channel's capacity in one batch.)

Reply:
True, the incoming requests are sent in tight loop here, so while it's being drained new incoming requests are being batched in the channel instead of being blocked by a full channel. The collecting list for batched requests does not have a capacity limit so we end up having more than the channel's capacity for one batch.

Since the operation we want to process is cluster reconfigure, which does not depend on individual cVM create request. I guess in the current tests, if every incoming request gets a response (thru the `done` channels in the tests), then we can be sure that all cVMs are added to the VM group